pkg/controller/cloud/nodecontroller.go

```diff
@@ -63,96 +74,358 @@ func NewCloudNodeController(
 	nodeInformer coreinformers.NodeInformer,
 	kubeClient clientset.Interface,
 	cloud cloudprovider.Interface,
-	nodeMonitorPeriod time.Duration) *CloudNodeController {
+	nodeMonitorPeriod time.Duration,
+	nodeStatusUpdateFrequency time.Duration) *CloudNodeController {
 
 	eventBroadcaster := record.NewBroadcaster()
 	recorder := eventBroadcaster.NewRecorder(api.Scheme, clientv1.EventSource{Component: "cloudcontrollermanager"})
 	eventBroadcaster.StartLogging(glog.Infof)
 	if kubeClient != nil {
 		glog.V(0).Infof("Sending events to api server.")
-		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.Core().RESTClient()).Events("")})
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
 	} else {
 		glog.V(0).Infof("No api server defined - no events will be sent to API server.")
 	}
 
 	cnc := &CloudNodeController{
-		nodeInformer:      nodeInformer,
-		kubeClient:        kubeClient,
-		recorder:          recorder,
-		cloud:             cloud,
-		nodeMonitorPeriod: nodeMonitorPeriod,
+		nodeInformer:              nodeInformer,
+		kubeClient:                kubeClient,
+		recorder:                  recorder,
+		cloud:                     cloud,
+		nodeMonitorPeriod:         nodeMonitorPeriod,
+		nodeStatusUpdateFrequency: nodeStatusUpdateFrequency,
 	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: cnc.AddCloudNode,
+	})
+
 	return cnc
 }
 
 // This controller deletes a node if kubelet is not reporting
 // and the node is gone from the cloud provider.
 func (cnc *CloudNodeController) Run() {
-	go func() {
-		defer utilruntime.HandleCrash()
+	defer utilruntime.HandleCrash()
 
-		go wait.Until(func() {
-			nodes, err := cnc.kubeClient.Core().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
-			if err != nil {
-				glog.Errorf("Error monitoring node status: %v", err)
+	// The following loops run communicate with the APIServer with a worst case complexity
+	// of O(num_nodes) per cycle. These functions are justified here because these events fire
+	// very infrequently. DO NOT MODIFY this to perform frequent operations.
+
+	// Start a loop to periodically update the node addresses obtained from the cloud
+	go wait.Until(cnc.UpdateNodeStatus, cnc.nodeStatusUpdateFrequency, wait.NeverStop)
+
+	// Start a loop to periodically check if any nodes have been deleted from cloudprovider
+	go wait.Until(cnc.MonitorNode, cnc.nodeMonitorPeriod, wait.NeverStop)
+}
+
+// UpdateNodeStatus updates the node status, such as node addresses
+func (cnc *CloudNodeController) UpdateNodeStatus() {
+	instances, ok := cnc.cloud.Instances()
```

Comment:
Ref. #45118 - we probably should make some decision on expected semantics of listing instances.